File: quine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int file_size(FILE *file)
{
    fseek(file, 0, SEEK_END);
    int size = ftell(file);
    fseek(file, 0, SEEK_SET);
    return size;
}

int split(char *buffer, char *delim, char ***output)
{
    char *ptr = buffer;
    int count = 0;
    int i = 0;
    while (ptr && *ptr)
    {
        ptr = strpbrk(ptr, delim);
        if (ptr)
            ptr += strspn(ptr, delim);
        count += 1;
    }
    *output = malloc(count * sizeof(char *));
    char *token = strtok(buffer, delim);
    while (token)
    {
        (*output)[i++] = token;
        token = strtok(NULL, delim);
    }
    return count;
}

void trim(char *ptr)
{
    while (*ptr == 0 || *ptr == ' ' || *ptr == '\n')
    {
        *ptr = 0;
        ptr--;
    }
}

char *readfile(char *filename)
{
    FILE *file = fopen(filename, "r");
    int size = file_size(file);
    char *buffer = malloc(size + 1);
    fread(buffer, size, 1, file);
    buffer[size] = 0;
    fclose(file);
    return buffer;
}

void writefile(char *filename, char *data)
{
    FILE *file = fopen(filename, "w");
    fputs(data, file);
    fclose(file);
}

typedef struct _name_list_node
{
    char *name;
    struct _name_list_node *next;
} name_list_node;

void help()
{
    char *data = readfile("help.txt");
    puts(data);
    free(data);
}

char **dict_get(char ***dict, int size, char *key)
{
    for (int i = 0; i < size; i++)
    {
        if (strcmp(dict[i][0], key) == 0)
            return dict[i];
    }
    return 0;
}

char *safecatc(char *buffer, int *size, char cat)
{
    int pos = strlen(buffer);
    if (pos + 1 >= *size)
    {
        buffer = realloc(buffer, *size + 4096);
        buffer[*size] = 0;
        *size += 4096;
    }
    buffer[pos] = cat;
    buffer[pos + 1] = 0;
    return buffer;
}

char *safecat(char *buffer, int *size, char *cat)
{
    if (strlen(buffer) + strlen(cat) >= size)
    {
        buffer = realloc(buffer, *size + 4096);
        buffer[*size] = 0;
        *size += 4096;
    }
    strcat(buffer, cat);
    return buffer;
}

char *render(char *template, char ***details, int size)
{
    int bsize = 4096;
    char *buffer = malloc(bsize);
    buffer[0] = 0;
    int i = 0;
    int len = strlen(template);
    while (i < len)
    {
        char c = template[i];
        if (c == '$')
        {
            int j = 1;
            while (template[i + j] != '$')
                j += 1;
            char *name = malloc(j);
            memcpy(name, &(template[i + 1]), j - 1);
            name[j - 1] = 0;
            char **item = dict_get(details, size, name);
            buffer = safecatc(buffer, &bsize, '"');
            buffer = safecat(buffer, &bsize, item[2]);
            buffer = safecat(buffer, &bsize, dict_get(details, size, item[5])[4]);
            buffer = safecatc(buffer, &bsize, '"');
            i += j;
        }
        else
        {
            buffer = safecatc(buffer, &bsize, c);
        }
        i += 1;
    }
    return buffer;
}

void build(char *argv[])
{
    char *buffer = readfile(argv[2]);
    char **lines;
    int ln = split(buffer, "\n", &lines);
    char ***langs = malloc(sizeof(char **) * ln);
    memset(langs, 0, sizeof(char **) * ln);
    int count = 0;
    for (int i = 0; i < ln; i++)
    {
        if (strlen(lines[i]) == 0)
            continue;
        char **parts;
        int pn = split(lines[i], "#", &parts);
        langs[count++] = parts;
    }
    free(lines);
    char ***details = malloc(sizeof(char **) * count);
    for (int i = 0; i < count; i++)
    {
        details[i] = malloc(sizeof(char *) * 6);
        details[i][0] = langs[i][0];
        char *cmd = malloc(strlen(langs[i][1]) + 10 + strlen(argv[2]) + 8 + 1);
        strcpy(cmd, langs[i][1]);
        strcat(cmd, " template ");
        strcat(cmd, argv[2]);
        strcat(cmd, " > otemp");
        system(cmd);
        free(cmd);
        details[i][1] = readfile("otemp");
        trim(details[i][1] + strlen(details[i][1]));
        system("rm otemp");
    }
    for (int i = 0; i < count; i++)
    {
        writefile("itemp", details[(i + 1) % count][1]);
        char *cmd = malloc(strlen(langs[i][1]) + 5 + 16 + 1);
        strcpy(cmd, langs[i][1]);
        strcat(cmd, " exec");
        strcat(cmd, " > otemp < itemp");
        system(cmd);
        free(cmd);
        char *output = readfile("otemp");
        system("rm itemp");
        system("rm otemp");
        char *ptr = strpbrk(output, "\n");
        *ptr++ = 0;
        trim(ptr + strlen(ptr));
        details[i][2] = output;
        details[i][3] = ptr;
        //
        writefile("itemp", details[i][3]);
        cmd = malloc(strlen(langs[(i - 1 + count) % count][1]) + 12 + 16 + 1);
        strcpy(cmd, langs[(i - 1 + count) % count][1]);
        strcat(cmd, " anti-escape");
        strcat(cmd, " > otemp < itemp");
        system(cmd);
        free(cmd);
        char *antied = readfile("otemp");
        system("rm itemp");
        system("rm otemp");
        trim(antied + strlen(antied));
        details[i][4] = antied;
        details[i][5] = details[(i + 1) % count][0];
    }
    char *code = render(details[0][1], details, count);
    FILE *codefile = fopen(argv[3], "w");
    fwrite(code, strlen(code), 1, codefile);
    free(code);
    fprintf(codefile, "%s\n", details[0][3]);
    fclose(codefile);
    printf("try run: cat %s", argv[3]);
    for (int i = 0; i < count; i++)
    {
        printf("|%s", langs[i][2]);
    }
    printf("|diff %s -\n", argv[3]);
    for (int i = 0; i < count; i++)
    {
        free(langs[i]);
        free(details[i][1]);
        free(details[i][2]);
        free(details[i][4]);
        free(details[i]);
    }
    free(langs);
    free(details);
    free(buffer);
}

int main(int argc, char *argv[])
{
    if (argc < 2)
        return -1;
    if (strcmp(argv[1], "template") == 0)
    {
        if (argc < 3)
            return -1;
        char *buffer = readfile(argv[2]);
        puts("#include<stdio.h>");
        char **lines;
        int ln = split(buffer, "\n", &lines);
        for (int i = 0; i < ln; i++)
        {
            if (strlen(lines[i]) == 0)
                continue;
            char **parts;
            int pn = split(lines[i], "#", &parts);
            printf("char*%s=$%s$;", parts[0], parts[0]);
            free(parts);
        }
        putchar('\n');
        free(lines);
        free(buffer);
    }
    else if (strcmp(argv[1], "exec") == 0)
    {
        char c;
        char *total_str = malloc(4096);
        char *cptr = total_str;
        while ((c = getchar()) != EOF)
        {
            *cptr++ = c;
        }
        *cptr = 0;
        trim(cptr);
        name_list_node head;
        name_list_node *tail = &head;
        int len = strlen(total_str);
        for (int i = 0; i < len; i++)
        {
            c = total_str[i];
            switch (c)
            {
            case '"':
                printf("%%c");
                tail->next = malloc(sizeof(name_list_node));
                tail = tail->next;
                tail->name = malloc(3);
                strcpy(tail->name, "34");
                tail->next = NULL;
                break;
            case '\n':
                printf("%%c");
                tail->next = malloc(sizeof(name_list_node));
                tail = tail->next;
                tail->name = malloc(3);
                strcpy(tail->name, "10");
                tail->next = NULL;
                break;
            case '$':
            {
                char *cache = malloc(64);
                strcpy(cache, "34,");
                char *ptr = &(cache[3]);
                int j = 1;
                while (total_str[i + j] != '$')
                {
                    *ptr++ = total_str[i + j];
                    j += 1;
                }
                i += j;
                strcpy(ptr, ",34");
                printf("%%c%%s%%c");
                tail->next = malloc(sizeof(name_list_node));
                tail = tail->next;
                tail->name = cache;
                tail->next = NULL;
            }
            break;
            default:
                putchar(c);
            }
        }
        putchar('\n');
        printf("int main(){printf(c");
        name_list_node *ptr = head.next;
        while (ptr != NULL)
        {
            printf(",%s", ptr->name);
            free(ptr->name);
            name_list_node *temp = ptr;
            ptr = ptr->next;
            free(temp);
        }
        puts(");putchar(10);}");
        free(total_str);
    }
    else if (strcmp(argv[1], "anti-escape") == 0)
    {
        char c;
        char *total_str = malloc(4096);
        char *cptr = total_str;
        while ((c = getchar()) != EOF)
        {
            *cptr++ = c;
        }
        *cptr = 0;
        trim(cptr);
        name_list_node head;
        name_list_node *tail = &head;
        int len = strlen(total_str);
        for (int i = 0; i < len; i++)
        {
            c = total_str[i];
            switch (c)
            {
            case '%':
                printf("%%%%");
                break;
            default:
                putchar(c);
            }
        }
        putchar('\n');
        free(total_str);
    }
    else if (strcmp(argv[1], "build") == 0)
    {
        build(argv);
    }
    else if (strcmp(argv[1], "help") == 0)
    {
        help();
    }
    else
    {
        puts("unkown opntions, use \"help\" to get help.");
        help();
    }
    return 0;
}

File: quine.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"strings"
)

func help() {
	data, _ := ioutil.ReadFile("help.txt")
	fmt.Println(string(data))
}

func render(template string, details map[string][]string) string {
	result := ""
	i := 0
	tempRune := []rune(template)
	for i < len(tempRune) {
		c := tempRune[i]
		if c == '$' {
			j := 1
			for tempRune[i+j] != '$' {
				j += 1
			}
			item := details[string(tempRune[i+1:i+j])]
			result += "\"" + item[1] + details[item[4]][3] + "\""
			i += j
		} else {
			result += string(c)
		}
		i += 1
	}
	return result
}

func build(filename string, codefilename string) {
	langs := [][]string{}
	details := [][]string{}
	buffer, _ := ioutil.ReadFile(os.Args[2])
	lines := strings.Split(string(buffer), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}
		langs = append(langs, strings.Split(line, "#"))
	}
	for _, lang := range langs {
		cmd := strings.Split(lang[1]+" template "+filename, " ")
		output, _ := exec.Command(cmd[0], cmd[1:]...).Output()
		temp := strings.TrimSpace(string(output))
		details = append(details, []string{lang[0], temp})
	}
	for i, _ := range langs {
		cmd := strings.Split(langs[i][1]+" exec", " ")
		proc := exec.Command(cmd[0], cmd[1:]...)
		stdin, _ := proc.StdinPipe()
		stdin.Write([]byte(details[(i+1)%len(langs)][1] + "\n"))
		stdin.Close()
		output, _ := proc.Output()
		lines = strings.Split(string(output), "\n")
		details[i] = append(details[i], lines[0])
		details[i] = append(details[i], strings.Join(lines[1:], "\n"))
		//
		antiCmd := strings.Split(langs[(i-1+len(langs))%len(langs)][1]+" anti-escape", " ")
		antiProc := exec.Command(antiCmd[0], antiCmd[1:]...)
		antiStdin, _ := antiProc.StdinPipe()
		antiStdin.Write([]byte(details[i][3] + "\n"))
		antiStdin.Close()
		antiOut, _ := antiProc.Output()
		details[i] = append(details[i], strings.TrimSpace(string(antiOut)))
		details[i] = append(details[i], details[(i+1)%len(langs)][0])
	}
	detailDict := make(map[string][]string)
	for _, d := range details {
		detailDict[d[0]] = d[1:]
	}
	code := render(details[0][1], detailDict) + details[0][3]
	ioutil.WriteFile(codefilename, []byte(code), 0755)
	fmt.Printf("try run: cat %s", codefilename)
	for _, lang := range langs {
		fmt.Print("|" + lang[2])
	}
	fmt.Printf("|diff %s -\n", codefilename)
}

func main() {
	if len(os.Args) < 2 {
		return
	}
	if os.Args[1] == "template" && len(os.Args) == 3 {
		buffer, _ := ioutil.ReadFile(os.Args[2])
		lines := strings.Split(string(buffer), "\n")
		fmt.Println("package main")
		fmt.Println("import \"fmt\"")
		fmt.Print("const(_qu=string(rune(34));_ln=string(rune(10))")
		for _, line := range lines {
			if line == "" {
				continue
			}
			name := strings.Split(line, "#")[0]
			fmt.Printf(";%s=$%s$", name, name)
		}
		fmt.Println(");")
	} else if os.Args[1] == "exec" {
		lines := []string{}
		scanner := bufio.NewScanner(os.Stdin)
		for scanner.Scan() {
			line := scanner.Text()
			lines = append(lines, line)
		}
		temp := []rune(strings.Join(lines, "\n"))
		// fmt.Println(lines)
		names := []string{}
		i := 0
		fmt.Print("")
		for i < len(temp) {
			c := temp[i]
			if c == '"' {
				fmt.Print("%s")
				names = append(names, "_qu")
			} else if c == '\n' {
				fmt.Print("%s")
				names = append(names, "_ln")
			} else if c == '$' {
				j := 1
				for temp[i+j] != '$' {
					j += 1
				}
				names = append(names, "_qu", string(temp[i+1:i+j]), "_qu")
				fmt.Print("%s%s%s")
				i += j
			} else {
				fmt.Print(string(c))
			}
			i += 1
		}
		fmt.Print("\n")
		fmt.Printf("func main(){fmt.Printf(golang,%s);fmt.Print(_ln)}\n", strings.Join(names, ","))
	} else if os.Args[1] == "anti-escape" {
		lines := []string{}
		scanner := bufio.NewScanner(os.Stdin)
		for scanner.Scan() {
			line := scanner.Text()
			lines = append(lines, line)
		}
		temp := strings.Join(lines, "\n")
		for _, c := range temp {
			if c == '%' {
				fmt.Print("%%")
			} else {
				fmt.Print(string(c))
			}
		}
	} else if os.Args[1] == "build" && len(os.Args) == 4 {
		build(os.Args[2], os.Args[3])
	} else if os.Args[1] == "help" {
		help()
	} else {
		fmt.Println("unkown opntions, use \"help\" to get help.")
		help()
	}
	return
}
